### ISSUE TITLE
Support for connect_cb

### ISSUE THREAD

**Issue by hallfox** [confluentinc/confluent-kafka-python#1097]
We have an experimental patch which adds support for setting `connect_cb` as a configuration, however looking back we found [this issue from 2017
](https://github.com/confluentinc/confluent-kafka-python/issues/219#issuecomment-315983155) which says that adding it would be hard.

And it is! By doing so we discovered a deadlock that would occur since the `connect_cb` is called in a different thread than the other callback configs due to in part the GIL being held during calls to `rd_kafka_assign`. Trying to solve this, we released the GIL for the duration of `rd_kafka_assign` and things seem to be working well now.

That context is here because I'm curious to know:
1. Is supporting `connect_cb` hard for other specific reasons?
2. Does the assign call need to hold the GIL for other safety reasons?

Basically I'm wondering if there are some invariants I may be stepping on with these changes. Thanks.

**Comment by edenhill** [comment]
What's your use of the connect_cb?

**Comment by hallfox** [comment]
We need to make a connection through a proxy address.

**Comment by edenhill** [comment]
To answer your original questions:

> Is supporting connect_cb hard for other specific reasons?

No, it should be straight forward as long as the callback doesn't do anything overly complex since it is called from an internal broker thread.

> Does the assign call need to hold the GIL for other safety reasons?

No, no need. But the deadlock may occur in many of the librdkafka C calls (any call that may need to aquire a broker lock), so all of them need to release the GIL if we were to add connect_cb support.
Before going down that route it would be good to understand the cost of releasing and re-acquiring the GIL. The underlying mutex itself is cheap, but maybe CPython does other things as well?

**Comment by hallfox** [comment]
> Before going down that route it would be good to understand the cost of releasing and re-acquiring the GIL.

So we did already try going down this route and have a patch basically ready to go, however we haven't done a broader analysis considering the performance impact of this. Given that this might be adding some fairly costly operations in a hot path, what do you think makes sense? Maybe releasing the GIL only under certain conditions?

**Comment by edenhill** [comment]
Since the librdkafka API does not make any guarantees what internal locks it may or may not acquire (and that would be a tough guarantee to maintain as the code evolves), we pretty much need to release the GIL on all rdkafka API calls to avoid a deadlock.

**Comment by hallfox** [comment]
We've already done this and have some anecdotal evidence that it works, should I open a PR and continue this discussion there?

**Comment by edenhill** [comment]
Please do, what I'm interest to know is a performance comparison for the hot-paths: produce(), poll(), and consume(), to see if there's a noticable difference in un-GILing on each call.
I guess poll() and consume() already do, so that leaves it to produce().

**Comment by MSeal** [comment]
Related PR needs to be evaluated for reuse with modern versions of the codebase and a test case added to validate before merge + linting resolutions
